clang/lib/FrontendTool/CMakeLists.txt

```diff
@@ -15,7 +15,9 @@ set(link_libs
 if(CLANG_ENABLE_CIR)
   list(APPEND link_libs
     clangCIRFrontendAction
+    MLIRIR
```

Comment:
Is it possible to sink this cmake change down to clangCIRFrontendAction?

Reply:
Good point. These dependencies are already there in clangCIRFrontendAction's CMakeLists.txt but it doesn't work for building Clang in shared library mode, where .so is linked separately and each of them needs their own static linking dependencies.

Reply:
hmmm, too bad, we probably need a more holistic solution at some point then!

For now, can you add both additions conditionally based on the shared library cmake variable?

Reply:
> For now, can you add both additions conditionally based on the shared library cmake variable?

I guess we can but that doesn't sound a common way. Literally `clangFrontendTool` is now depending on MLIRIR regardless of `clangCIRFrontendAction` depends on it or not. Dependencies should be explicitly specified anyways no matter what the linking type is.

Is your concern at this point about polluting the non-CIR code path that may make it harder to do merge integration? I feel this might be inevitable as we work deeper and broader on this project. The Meta llvm team uses a special comment marker to ease that which we may consider, but still merge conflicts are all over the places.

Reply:
> I guess we can but that doesn't sound a common way. Literally `clangFrontendTool` is now depending on MLIRIR regardless of `clangCIRFrontendAction` depends on it or not. Dependencies should be explicitly specified anyways no matter what the linking type is.

Not sure I agree, it's pretty common practice to have a dependency tree between libraries, different Clang/LLVM parts don't specify stuff over and over again to work with shared libraries, do they? At least I haven't noticed this pattern you're mentioning, but I might be wrong. @lanza do you have an opinion here?

> Is your concern at this point about polluting the non-CIR code path that may make it harder to do merge integration? 

Not for this specific change, my concern here is about CMAKE. It's easy to make hacky changes to get stuff working (I've done many) but that are hiding a bigger lack of proper solution because cmake can get tricky.

Anyways, let's move on with this and we can sort this out post-commit. Thanks for working on this, LGTM

Reply:
>  different Clang/LLVM parts don't specify stuff over and over again to work with shared libraries, do they? 

It's common to specify direct dependencies explicitly for a module, instead of relying on hidden indirect dependencies. E.g, Clang.exe relies on LLVM IPO and Vectorize (https://github.com/llvm/clangir/blob/a98cafd1d0f1144bc9149f7781de41c1abf7c960/clang/tools/driver/CMakeLists.txt#L6), and LLVM IP relies on Vectorize too (https://github.com/llvm/clangir/blob/700cd99061edeeba7b657e32acca940225fa25ae/llvm/lib/Transforms/IPO/CMakeLists.txt#L72).

I think the way it works is that when building in shared mode, LLVMVectorize.so appears on the the linking command line for both Clang.exe and LLVMIPO.so. When building in non-shared mode,  LLVMVectorize.a and LLVMIPO.a appear only once on the Clang.exe linking command line. The build system should be able to avoid duplicating LLVMVectorize.a there.

Another example is that clangFrontendTool has an explicit dependency on both clangBasic and clangCodeGen, while clangCodeGen also depends on clangBasic.